### DISCUSSION TITLE
How to compile function with `impl diesel::Connection`?

### DISCUSSION THREAD

**Discussion by gabyx** [diesel-rs/diesel#3909] [Q&A]
I am trying to build a `transaction` function `insert_job` which looks like the following:
[Example](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KW2RlcGVuZGVuY2llcy51dWlkXQp2ZXJzaW9uID0gIjEuNi4xIgpmZWF0dXJlcyA9IFsidjQiLCAiZmFzdC1ybmciLCAibWFjcm8tZGlhZ25vc3RpY3MiLCAic2VyZGUiXQoKW2RlcGVuZGVuY2llcy5kaWVzZWxdCnZlcnNpb24gPSAiMi4xLjAiCmZlYXR1cmVzID0gWyJwb3N0Z3JlcyIsICJ1dWlkIl0KW2RlcGVuZGVuY2llcy5kaWVzZWwtZGVyaXZlLWVudW1dCnZlcnNpb24gPSAiMi4xLjAiCmZlYXR1cmVzID0gIFsicG9zdGdyZXMiXQoqLwoKCgpwdWIgbW9kIHNxbF90eXBlcyB7CiAgICAjW2Rlcml2ZShkaWVzZWw6OnF1ZXJ5X2J1aWxkZXI6OlF1ZXJ5SWQsIGRpZXNlbDo6c3FsX3R5cGVzOjpTcWxUeXBlKV0KICAgICNbZGllc2VsKHBvc3RncmVzX3R5cGUobmFtZSA9ICJyZXN1bHRfdCIpKV0KICAgIHB1YiBzdHJ1Y3QgUmVzdWx0VDsKCiAgICAjW2Rlcml2ZShkaWVzZWw6OnF1ZXJ5X2J1aWxkZXI6OlF1ZXJ5SWQsIGRpZXNlbDo6c3FsX3R5cGVzOjpTcWxUeXBlKV0KICAgICNbZGllc2VsKHBvc3RncmVzX3R5cGUobmFtZSA9ICJzdGF0dXNfdCIpKV0KICAgIHB1YiBzdHJ1Y3QgU3RhdHVzVDsKfQoKZGllc2VsOjp0YWJsZSEgewogICAgdXNlIGRpZXNlbDo6c3FsX3R5cGVzOjoqOwogICAgdXNlIHN1cGVyOjpzcWxfdHlwZXM6OlN0YXR1c1Q7CiAgICB1c2Ugc3VwZXI6OnNxbF90eXBlczo6UmVzdWx0VDsKCiAgICBqb2JzIChpZCkgewogICAgICAgIGlkIC0+IFV1aWQsCiAgICAgICAgc3RhdHVzIC0+IFN0YXR1c1QsCiAgICAgICAgYmxvYl9kaWdlc3QgLT4gVmFyY2hhciwKICAgICAgICBuYW1lIC0+IFZhcmNoYXIsCiAgICAgICAgY29udmVydGVyX3Jlc3VsdCAtPiBSZXN1bHRULAogICAgICAgIGNvbnZlcnRlcl9sb2cgLT4gVGV4dCwKICAgIH0KfQp1c2UgZGllc2VsOjpwcmVsdWRlOjoqOwp1c2UgZGllc2VsX2Rlcml2ZV9lbnVtOwp1c2UgdXVpZDo6VXVpZDsKCi8vIENSRUFURSBUWVBFIEpvYlN0YXR1cyBhcyBFTlVNKC4uLik7awojW2Rlcml2ZShDbG9uZSwgQ29weSwgRGVidWcsIEhhc2gsIFBhcnRpYWxFcSwgRXEsIGRpZXNlbF9kZXJpdmVfZW51bTo6RGJFbnVtKV0KI1tFeGlzdGluZ1R5cGVQYXRoID0gInNxbF90eXBlczo6U3RhdHVzVCJdCnB1YiBlbnVtIEpvYlN0YXR1cyB7CiAgICBJZGxlLAogICAgUXVldWVkLAogICAgUnVubmluZywKICAgIERvbmUsCn0KCiNbZGVyaXZlKENsb25lLCBDb3B5LCBEZWJ1ZywgSGFzaCwgUGFydGlhbEVxLCBFcSwgZGllc2VsX2Rlcml2ZV9lbnVtOjpEYkVudW0pXQojW0V4aXN0aW5nVHlwZVBhdGggPSAic3FsX3R5cGVzOjpSZXN1bHRUIl0KcHViIGVudW0gQ29udmVydGVyUmVzdWx0IHsKICAgIE5vbmUsCiAgICBGYWlsdXJlLAogICAgU3VjY2VzcywKfQoKI1tkZXJpdmUoUXVlcnlhYmxlLCBTZWxlY3RhYmxlLCBJbnNlcnRhYmxlKV0KI1tkaWVzZWwodGFibGVfbmFtZSA9IGpvYnMpXQojW2RpZXNlbChjaGVja19mb3JfYmFja2VuZChkaWVzZWw6OnBnOjpQZykpXQpwdWIgc3RydWN0IERiSm9iIHsKICAgIHB1YiBpZDogVXVpZCwKICAgIHB1YiBzdGF0dXM6IEpvYlN0YXR1cywKCiAgICBwdWIgYmxvYl9kaWdlc3Q6IFN0cmluZywKCiAgICBwdWIgbmFtZTogU3RyaW5nLAoKICAgIHB1YiBjb252ZXJ0ZXJfcmVzdWx0OiBDb252ZXJ0ZXJSZXN1bHQsCiAgICBwdWIgY29udmVydGVyX2xvZzogU3RyaW5nLAp9CgpmbiBpbnNlcnRfam9iPEM+KGNvbm46ICZtdXQgQykKd2hlcmUKICAgIEM6IENvbm5lY3Rpb24sCnsKICAgIHVzZSBkaWVzZWw6OntpbnNlcnRfaW50bywgQ29ubmVjdGlvbiwgUnVuUXVlcnlEc2x9OwogICAgdXNlIGNyYXRlOjpqb2JzOjpkc2w6Oio7CiAgICAKICAgIGxldCBqID0gRGJKb2IgewogICAgICAgIGlkOiBVdWlkOjpuZXdfdjQoKSwKICAgICAgICBzdGF0dXM6IEpvYlN0YXR1czo6SWRsZSwKICAgICAgICBibG9iX2RpZ2VzdDogImFzZGYiLnRvX293bmVkKCksCiAgICAgICAgbmFtZTogImFzZGYiLmNsb25lKCksCiAgICAgICAgY29udmVydGVyX3Jlc3VsdDogQ29udmVydGVyUmVzdWx0OjpOb25lLAogICAgICAgIGNvbnZlcnRlcl9sb2c6ICIiLnRvX293bmVkKCksCiAgICB9OwoKICAgIGluc2VydF9pbnRvKGpvYnMpLnZhbHVlcyhqKS5leGVjdXRlKGNvbm4pOwp9CgpmbiBtYWluKCkgewogICAgCn0=)
```rust
use diesel::{insert_into, Connection, RunQueryDsl};

use super::{
    models::{ConverterResult, DbJob, JobStatus},
    schema::jobs::dsl::*,
};
use crate::job::JobBundle;

fn insert_job<C>(conn: &mut C, job: &JobBundle)
where
    C: Connection,
{
    let j = DbJob {
        id: job.id,
        status: JobStatus::Idle,
        blob_digest: job.blob_digest.clone(),
        name: job.name.clone(),
        converter_result: ConverterResult::None,
        converter_log: "".to_owned(),
    };

    insert_into(jobs).values(j).execute(conn);
}
```

[The `execute(conn)` actually is excatly](https://docs.rs/diesel/latest/diesel/prelude/trait.RunQueryDsl.html#method.execute) the same written as I did but Rust wants me to put all sorts of bounds on the `Connection` (why?):

```rust
error[E0277]: the trait bound `<C as diesel::Connection>::Backend: DieselReserveSpecialization` is not satisfied
    --> components/common/src/db/transactions.rs:21:41
     |
21   |     insert_into(jobs).values(j).execute(conn);
     |                                 ------- ^^^^ the trait `DieselReserveSpecialization` is not implemented for `<C as diesel::Connection>::Backend`
     |                                 |
     |                                 required by a bound introduced by this call
     |
     = note: required for `InsertStatement<table, ValuesClause<..., ...>>` to implement `QueryFragment<<C as diesel::Connection>::Backend>`
     = note: the full type name has been written to '/home/nixos/Desktop/Repos/markdown-to-pdf-service/target/debug/deps/common-407478c3fbb79c9d.long-type-15797717890643199462.txt'
     = note: required for `InsertStatement<table, ValuesClause<..., ...>>` to implement `ExecuteDsl<C>`
     = note: the full type name has been written to '/home/nixos/Desktop/Repos/markdown-to-pdf-service/target/debug/deps/common-407478c3fbb79c9d.long-type-15797717890643199462.txt'
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> /home/nixos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/diesel-2.1.4/src/query_dsl/mod.rs:1432:15
     |
1429 |     fn execute(self, conn: &mut Conn) -> QueryResult<usize>
     |        ------- required by a bound in this associated function
...
1432 |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::execute`
help: consider further restricting the associated type
     |
11   |     C: Connection, <C as diesel::Connection>::Backend: DieselReserveSpecialization
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0277]: the trait bound `DefaultValues: QueryFragment<<C as diesel::Connection>::Backend, <<C as diesel::Connection>::Backend as SqlDialect>::DefaultValueClauseForInsert>` is not satisfied
    --> components/common/src/db/transactions.rs:21:41
 ```
and the list goes on and on...

Does anybody know where I have to watch out when writting such a generic function. I thougth that works, but it seems totally nontrivial.

[Example](https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KW2RlcGVuZGVuY2llcy51dWlkXQp2ZXJzaW9uID0gIjEuNi4xIgpmZWF0dXJlcyA9IFsidjQiLCAiZmFzdC1ybmciLCAibWFjcm8tZGlhZ25vc3RpY3MiLCAic2VyZGUiXQoKW2RlcGVuZGVuY2llcy5kaWVzZWxdCnZlcnNpb24gPSAiMi4xLjAiCmZlYXR1cmVzID0gWyJwb3N0Z3JlcyIsICJ1dWlkIl0KW2RlcGVuZGVuY2llcy5kaWVzZWwtZGVyaXZlLWVudW1dCnZlcnNpb24gPSAiMi4xLjAiCmZlYXR1cmVzID0gIFsicG9zdGdyZXMiXQoqLwoKCgpwdWIgbW9kIHNxbF90eXBlcyB7CiAgICAjW2Rlcml2ZShkaWVzZWw6OnF1ZXJ5X2J1aWxkZXI6OlF1ZXJ5SWQsIGRpZXNlbDo6c3FsX3R5cGVzOjpTcWxUeXBlKV0KICAgICNbZGllc2VsKHBvc3RncmVzX3R5cGUobmFtZSA9ICJyZXN1bHRfdCIpKV0KICAgIHB1YiBzdHJ1Y3QgUmVzdWx0VDsKCiAgICAjW2Rlcml2ZShkaWVzZWw6OnF1ZXJ5X2J1aWxkZXI6OlF1ZXJ5SWQsIGRpZXNlbDo6c3FsX3R5cGVzOjpTcWxUeXBlKV0KICAgICNbZGllc2VsKHBvc3RncmVzX3R5cGUobmFtZSA9ICJzdGF0dXNfdCIpKV0KICAgIHB1YiBzdHJ1Y3QgU3RhdHVzVDsKfQoKZGllc2VsOjp0YWJsZSEgewogICAgdXNlIGRpZXNlbDo6c3FsX3R5cGVzOjoqOwogICAgdXNlIHN1cGVyOjpzcWxfdHlwZXM6OlN0YXR1c1Q7CiAgICB1c2Ugc3VwZXI6OnNxbF90eXBlczo6UmVzdWx0VDsKCiAgICBqb2JzIChpZCkgewogICAgICAgIGlkIC0+IFV1aWQsCiAgICAgICAgc3RhdHVzIC0+IFN0YXR1c1QsCiAgICAgICAgYmxvYl9kaWdlc3QgLT4gVmFyY2hhciwKICAgICAgICBuYW1lIC0+IFZhcmNoYXIsCiAgICAgICAgY29udmVydGVyX3Jlc3VsdCAtPiBSZXN1bHRULAogICAgICAgIGNvbnZlcnRlcl9sb2cgLT4gVGV4dCwKICAgIH0KfQp1c2UgZGllc2VsOjpwcmVsdWRlOjoqOwp1c2UgZGllc2VsX2Rlcml2ZV9lbnVtOwp1c2UgdXVpZDo6VXVpZDsKCi8vIENSRUFURSBUWVBFIEpvYlN0YXR1cyBhcyBFTlVNKC4uLik7awojW2Rlcml2ZShDbG9uZSwgQ29weSwgRGVidWcsIEhhc2gsIFBhcnRpYWxFcSwgRXEsIGRpZXNlbF9kZXJpdmVfZW51bTo6RGJFbnVtKV0KI1tFeGlzdGluZ1R5cGVQYXRoID0gInNxbF90eXBlczo6U3RhdHVzVCJdCnB1YiBlbnVtIEpvYlN0YXR1cyB7CiAgICBJZGxlLAogICAgUXVldWVkLAogICAgUnVubmluZywKICAgIERvbmUsCn0KCiNbZGVyaXZlKENsb25lLCBDb3B5LCBEZWJ1ZywgSGFzaCwgUGFydGlhbEVxLCBFcSwgZGllc2VsX2Rlcml2ZV9lbnVtOjpEYkVudW0pXQojW0V4aXN0aW5nVHlwZVBhdGggPSAic3FsX3R5cGVzOjpSZXN1bHRUIl0KcHViIGVudW0gQ29udmVydGVyUmVzdWx0IHsKICAgIE5vbmUsCiAgICBGYWlsdXJlLAogICAgU3VjY2VzcywKfQoKI1tkZXJpdmUoUXVlcnlhYmxlLCBTZWxlY3RhYmxlLCBJbnNlcnRhYmxlKV0KI1tkaWVzZWwodGFibGVfbmFtZSA9IGpvYnMpXQojW2RpZXNlbChjaGVja19mb3JfYmFja2VuZChkaWVzZWw6OnBnOjpQZykpXQpwdWIgc3RydWN0IERiSm9iIHsKICAgIHB1YiBpZDogVXVpZCwKICAgIHB1YiBzdGF0dXM6IEpvYlN0YXR1cywKCiAgICBwdWIgYmxvYl9kaWdlc3Q6IFN0cmluZywKCiAgICBwdWIgbmFtZTogU3RyaW5nLAoKICAgIHB1YiBjb252ZXJ0ZXJfcmVzdWx0OiBDb252ZXJ0ZXJSZXN1bHQsCiAgICBwdWIgY29udmVydGVyX2xvZzogU3RyaW5nLAp9CgpmbiBpbnNlcnRfam9iPEM+KGNvbm46ICZtdXQgQykKd2hlcmUKICAgIEM6IENvbm5lY3Rpb24sCnsKICAgIHVzZSBkaWVzZWw6OntpbnNlcnRfaW50bywgQ29ubmVjdGlvbiwgUnVuUXVlcnlEc2x9OwogICAgdXNlIGNyYXRlOjpqb2JzOjpkc2w6Oio7CiAgICAKICAgIGxldCBqID0gRGJKb2IgewogICAgICAgIGlkOiBVdWlkOjpuZXdfdjQoKSwKICAgICAgICBzdGF0dXM6IEpvYlN0YXR1czo6SWRsZSwKICAgICAgICBibG9iX2RpZ2VzdDogImFzZGYiLnRvX293bmVkKCksCiAgICAgICAgbmFtZTogImFzZGYiLmNsb25lKCksCiAgICAgICAgY29udmVydGVyX3Jlc3VsdDogQ29udmVydGVyUmVzdWx0OjpOb25lLAogICAgICAgIGNvbnZlcnRlcl9sb2c6ICIiLnRvX293bmVkKCksCiAgICB9OwoKICAgIGluc2VydF9pbnRvKGpvYnMpLnZhbHVlcyhqKS5leGVjdXRlKGNvbm4pOwp9CgpmbiBtYWluKCkgewogICAgCn0=)

**Comment by weiznich** [comment]
`QueryDsl::execute` has a additional trait bound that is not satisfied for your generic code. The documentation linked in your post states that it requires `Self: ExecuteDsl<Conn>`, where `Self` refers to whatever type you call that method on. Now rustc tries to be helpful and ties to see which other traits you could require to fulfill that bound. In this case it does not end up being helpful at all. I generally advice people not to write that kind of code at all, given these issues.

The reasoning for requiring these bounds is that if you want to check things at compile time you need to provide this information at compile time. Therefore if you want to have a generic compile time checked function you need to prove that all possible cases hold the required invariants.

**Reply by gabyx** [reply]
Thanks for the explanation:

The follow up I have is: How is it then possible to accept a `conn: &mut diesel::PgConnection` and it works, but casting to the trait `Conn` then does not satisfy `<type>.execute(&mut conn)` where `<type>` is this `InsertStatement<...>` type?

Is my approach really not solvable?

**Reply by gabyx** [reply]
The full type is 

```rust
InsertStatement<table, query_builder::insert_statement::ValuesClause<(DefaultableColumnInsertValue<ColumnInsertValue<columns::id, diesel::expression::bound::Bound<diesel::sql_types::Uuid, uuid::Uuid>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::status, diesel::expression::bound::Bound<StatusT, models::JobStatus>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::blob_digest, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::name, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::converter_result, diesel::expression::bound::Bound<ResultT, models::ConverterResult>>>, DefaultableColumnInsertValue<ColumnInsertValue<columns::converter_log, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>>), table>>
```

**Reply by weiznich** [reply]
> : How is it then possible to accept a conn: &mut diesel::PgConnection and it works, but casting to the trait Conn then does not satisfy <type>.execute(&mut conn) where <type> is this InsertStatement<...> type?

That's essentially already answered above. Rustc can prove for a concrete type that all the bounds are satisfied. It cannot prove that for any possible type that just implements `Connection` as it might miss quite these other bounds listed in the error message. 
The critical part here is mostly not the connection type itself, but the backend type attached to the connection. That's hidden behind the `Connection::Backend` associated type.

> Is my approach really not solvable?

It's solvable, by adding a lot of trait bounds. I just advice not doing that and I won't write the code for that for other people (as it does not scale).

**Reply by gabyx** [reply]
Ah so as I come from C++, in rust the compiler wants to satisfy right at the invocation that some traits are fullfilled for the associated type `Backend`. In C++ the compiler would just throw that thing into `execute` and it will work. The connection is a PgConnection after all. I am just wondering what the type is, on which I could abstract a Connection and simply hand that through the application (not a concrete type `PgConnection`), basically whats the interface for `PgConnection` which I can use with queries. Probably that does not exist, as of the heavy use of generics in the libaray itself?

**Reply by weiznich** [reply]
Yes rust generics work completely different than c++ templates. 

> I am just wondering what the type is, on which I could abstract a Connection and simply hand that through the application (not a concrete type PgConnection), basically whats the interface for PgConnection which I can use with queries. Probably that does not exist, as of the heavy use of generics in the libaray itself?

Well it depends on what you are trying to do. If you want to abstract over different database connection types for the **same** backend, i.e only such that target postgresql, you can easily do that by using `impl Connection<Backend = Pg>` as argument. That will just work. 
What will not work is using that to abstract over different database backend (i.e. postgresql and sqlite) as those backends support different sql dialects. To allow using the database specific extensions and enforcing that this is done correctly diesel requires this information to be present at compile time. If you ever want to abstract over different backends you should look into the documentation of `#[derive(MultiConnection)]` which lets you select which backends should be supported.

**Reply by gabyx** [reply]
Ah ok, thanks thats basically the answer I was looking for. Thanks alot!
